lib/sensors/accelerometer.cpp

```diff
@@ -2,17 +2,83 @@
+    logger.log(core::LogLevel::kFatal, "Failed to read the accelerometer device ID");
+    return std::nullopt;
+  }
+  if (*device_id != kExpectedDeviceIdValue) {
+    logger.log(core::LogLevel::kFatal, "Failure, mismatched device ID for accelerometer");
+    return std::nullopt;
+  }
+  const core::Result ctrl1_result
+    = i2c->writeByteToRegister(device_address, kCtrl1Address, kCtrl1Value);
+  if (ctrl1_result == core::Result::kFailure) { return std::nullopt; };
+  const core::Result ctrl2_result
+    = i2c->writeByteToRegister(device_address, kCtrl2Address, kCtrl2Value);
+  if (ctrl2_result == core::Result::kFailure) { return std::nullopt; };
+  const core::Result ctrl6_result
+    = i2c->writeByteToRegister(device_address, kCtrl6Address, kCtrl6Value);
+  if (ctrl6_result == core::Result::kFailure) { return std::nullopt; };
+  return Accelerometer(logger, i2c, channel, device_address);
+}
+
+Accelerometer::Accelerometer(core::ILogger &logger,
+                             std::shared_ptr<io::II2c> i2c,
+                             const std::uint8_t channel,
+                             const std::uint8_t device_address)
     : logger_(logger),
       i2c_(i2c),
-      channel_(channel)
+      channel_(channel),
+      device_address_(device_address)
 {
 }
 
 Accelerometer::~Accelerometer()
 {
 }
 
+std::optional<core::Result> Accelerometer::isValueReady()
+{
+  // check to see if the values are ready to be read
+  const auto data_ready = i2c_->readByte(device_address_, kDataReady);
+  if (!data_ready) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read acceleration data");
+    return std::nullopt;
+  }
+  if (*data_ready % 2 == 0) {
+    logger_.log(core::LogLevel::kWarn, "Failed to read acceleration data as it is not ready");
+    return core::Result::kFailure;
+  }
+  return core::Result::kSuccess;
+}
+
+// TODOLater: current settings of the accelerometer make it read in +-16g but with high noise. Check
+// whether other configuration could be better
+std::optional<core::RawAccelerationData> Accelerometer::read()
+{
+  const auto result_x = getRawAcceleration(core::Axis::kX);
+  if (!result_x) { return std::nullopt; }
+  const std::int32_t x_acceleration = getAccelerationFromRawValue(*result_x);
+  const auto result_y               = getRawAcceleration(core::Axis::kY);
+  if (!result_y) { return std::nullopt; }
+  const std::int32_t y_acceleration = getAccelerationFromRawValue(*result_y);
+  const auto result_z               = getRawAcceleration(core::Axis::kZ);
+  if (!result_z) { return std::nullopt; }
+  const std::int32_t z_acceleration = getAccelerationFromRawValue(*result_z);
+  core::WallClock time;
+  const std::optional<core::RawAccelerationData> acceleration_3_axis{
+    std::in_place, x_acceleration, y_acceleration, z_acceleration, time.now()};
+  logger_.log(core::LogLevel::kDebug, "Successfully read accelerometer data");
+  return acceleration_3_axis;
+}
+
 std::uint8_t Accelerometer::getChannel() const
```

Comment:
when would you need to get the channel of an accelerometer?

also for my info what is the channel? is it the i2c bus that the accelerometer is on?

Reply:
Mux

Reply:
oh ok makes sense